### ISSUE TITLE
Linear Multiple Scale Methods

### ISSUE THREAD

**Issue by qtw1998** [mnslarcher/kmeans-anchors-ratios#3]
Hi sir! Thank you super much for your contributions!!
Besides I've one another idea for the anchor settings:

1. Use Kmeans++ to get the anchor size on the default size (1280 * 720):
   ```python
    kmeans3 = KMeans(n_clusters=9)
    kmeans3.fit(x)
    y_kmeans3 = kmeans3.predict(x)
    # centers3 = kmeans3.cluster_centers_

    yolo_anchor_average=[]
    for ind in range (9):
        yolo_anchor_average.append(np.mean(x[y_kmeans3==ind],axis=0))
    yolo_anchor_average=np.array(yolo_anchor_average)

    yolo_anchor_average=np.array(yolo_anchor_average)

    plt.scatter(x[:, 0], x[:, 1], c=y_kmeans3, s=2, cmap='viridis')
    plt.scatter(yolo_anchor_average[:, 0], yolo_anchor_average[:, 1], c='red', s=50);
    print("yolo_anchor_average:", yolo_anchor_average)
   ```
and the results are 
```
array([[ 18.53132741,  18.6386687 ],
       [250.02347036, 169.41983063],
       [160.17930258, 105.73356663],
       [511.30567041, 436.80050207],
       [ 91.88059645,  64.4406237 ],
       [ 42.2448986 ,  44.88746252],
       [234.15546485, 334.40798186],
       [ 75.79529888, 160.21260118],
       [384.29831721, 257.61720803]])
```

2. The original width height ratio is 16: 9, so I put it in 1280 * 1280 and then rescale it into 640 * 640 (efficientdet-d1) ,and then get the followings: `<MAT above> / 1280`
```array([[0.0144776 , 0.01456146],
       [0.19533084, 0.13235924],
       [0.12514008, 0.08260435],
       [0.39945756, 0.34125039],
       [0.07178172, 0.05034424],
       [0.03300383, 0.03506833],
       [0.18293396, 0.26125624],
       [0.05921508, 0.12516609],
       [0.30023306, 0.20126344]])
```
My thinking is like: 
![image](https://user-images.githubusercontent.com/34357479/82636079-64b46d80-9c34-11ea-9378-ee46301a56f3.png)
3. Then I changed the codes like:
```python
for stride in self.strides:
            boxes_level = []
            base_anchor_size = self.anchor_scale * stride
            kmeans_anchors = [[0.0144776 , 0.01456146],
                                [0.19533084, 0.13235924],
                                [0.12514008, 0.08260435],
                                [0.39945756, 0.34125039],
                                [0.07178172, 0.05034424],
                                [0.03300383, 0.03506833],
                                [0.18293396, 0.26125624],
                                [0.05921508, 0.12516609],
                                [0.40023306, 0.60126344]]
    
            for als in kmeans_anchors: # als: anchor_linear_scale
                anchor_size_x_2 = base_anchor_size * als[0] * 2
                anchor_size_y_2 = base_anchor_size * als[1] * 2

                x = np.arange(stride / 2, image_shape[1], stride)
                y = np.arange(stride / 2, image_shape[0], stride)
                xv, yv = np.meshgrid(x, y)
                xv = xv.reshape(-1)
                yv = yv.reshape(-1)

                # y1,x1,y2,x2
                boxes = np.vstack((yv - anchor_size_y_2, xv - anchor_size_x_2,
                                   yv + anchor_size_y_2, xv + anchor_size_x_2))
                boxes = np.swapaxes(boxes, 0, 1)
                boxes_level.append(np.expand_dims(boxes, axis=1))
            # concat anchors on the same level to the reshape NxAx4
            boxes_level = np.concatenate(boxes_level, axis=1)
            boxes_all.append(boxes_level.reshape([-1, 4]))
```
Finally the inferring results tell me I'm wrong actually...
![image](https://user-images.githubusercontent.com/34357479/82636128-8f062b00-9c34-11ea-80a3-58c031faabca.png)
Now I'm really hoping to get your hleping hands.. Thx bro if you see this issue

**Comment by mnslarcher** [comment]
Hi @qtw1998 , thanks for the appreciation!

Some suggestions based on my current understanding:
* An important factor of the K-Means of this repository(and the original one) is that it uses the IoU as a measure of distance. This is something that standard K-Means implementations such as sklearn.cluster.KMeans do not currently support. Does your KMeans implementation use IoU as distance measurement?
* I think is not a good idea to use the "raw" boxes as input of K-Means. Infact if you see the code of this repo I run K-Means only after normalizing the boxes so they have an area of 1. If you don't do this, you K-Means will be driven mainly by the size of the boxes and not their shape that in my opinion in most of the cases is the most important factor (except if you have very small objects like another guy that opened an issue here)
* I work on this only during the weekend but for sure I'll add the code for bringing back to the original size the anchors, maybe that can help you for your implementation
* If you see in the code I put an option to run more than one time K-Means. Most of the time the result is that all the runs converge to the same or almost the same centroids, probably the K-Means initialization is not so important for this problem with few (2) dimensions
